自動更新機能の実装

YamamotoMao35 opened the pull request:
# WHAT
自動更新機能の実装

# WHY
自動更新機能を実装することで、ユーザーが投稿したコメントを自動で取得して別のユーザーの画面に表示させることができる。

https://gyazo.com/6b9f722319d32b461bb8f658320b4ff1

Commit:
Update _message.html.haml

Update _message.html.haml

表示されているメッセージのHTMLのタグ内のカスタムデータ属性を入れてid情報を追加した。

Commit:
Create messages_controller.rb

Create messages_controller.rb

controllersディレクトリの直下にapiディレクトリを作成し、
コントローラーを作成した。

Commit:
Update messages_controller.rb

Update messages_controller.rb

indexアクションの中で送られてきたメッセージのidを取得して
それよりあとに投稿されたメッセージをDBから取得するようにした。

Commit:
Update routes.rb

Update routes.rb

コードを追加し、apiディレクトリのmessagesコントローラーのindexアクションが動いて、json形式でレスポンスするように指定した。

Commit:
Create index.json.jbuilder

Create index.json.jbuilder

viewsディレクトリ直下にもapiディレクトリを作成し、ファイルを作成することでjson形式で取得した値を配列でレスポンスできるようにした。

Commit:
Update jbuilder files

Update jbuilder files

jbuilderファイルの記述を修正した。
createの方は記述を少なくし、indexの方はformat_posted_timeというヘルパーメソッドが定義されていないため行った

Commit:
Update message_controller.rb

Update message_controller.rb

ファイル内の記述を修正した。
今回は検索ではないのでLIKE句は使わない

Commit:
Update message.js

Update message.js

imageをimage.urlとすることで画像のurlだけを取得するようにし、
作成したindexアクションを動かすリクエストを実装した。

Commit:
Update message.js

Update message.js

HTMLを組み立てる関数を作成し、取得した最新のメッセージをブラウザのメッセージ一覧に追加できるようにした。

Commit:
Modify files description

Modify files description

ファイル内の誤記を修正した。

Commit:
Update index.html.haml

Update index.html.haml

group_idについてカスタムデータ属性を定義して、情報としてgroupのidを指定した。

Commit:
Update messages_controller.rb

Update messages_controller.rb

検索条件をグループごとにすることでグループごとで最新の値を取得できるようにした。

Commit:
Update message.js

Update message.js

定義していたbuildMessageHTMLを使わずに、先に定義していたbuildHTMLを72行目のreloadMessagesへ渡すことで行数を減らした。
73行目で変数group_idを定義し、last_message_idについてはカスタムデータ属性の値を取得するためvalue()からdata()とした。
また設定したgroup_idを79行目のハッシュの中に入れた。
さらにif (messages.length != 0)とすることでメッセージが入った時だけスクロールするようにした。
94行目でsetInterval関数を使うことで自動更新を実現した。

Commit:
Update message.js

Update message.js

メッセージ画面のみ自動更新するようにし、変数の間違いを修正した。

TOMARUMARU commented on app/assets/javascripts/message.js:
```diff
@@ -1,7 +1,7 @@
   function buildHTML(message){
-    if ( message.image ) {
-      var html = `<div class="message">
+    if ( message.body && message.image.url) {
```
```
<p class="message__text">
    ${message.body}
 </p>
<img src="${message.image.url}">
```
この記述以外共通しているので、
body, imageそれぞれある時は↑の記述を、ない時は空の文字列を変数に入れて、変数を出力するようにしましょう

YamamotoMao35 replied:
修正しました。

Commit:
Update message.js

Update message.js

bodyを表示する箇所とimageを表示する箇所をそれぞれ変数bodyHTMLとimageHTMLとし、その中に三項演算子で条件をつけ、変数htmlの箇所に入れた。
そのおかげで記述が短くなった。
実際にメッセージ画面で予定通りの動作をすることも確認した。

TOMARUMARU reviewed (approved):
LGTM
